x/blockdb/database.go

```diff
@@ -0,0 +1,1319 @@
+			zap.Int("recoveredBlocks", len(recoveredHeights)),
+			zap.Uint64("finalNextWriteOffset", s.nextDataWriteOffset.Load()),
+			zap.Uint64("maxContiguousBlockHeight", heights.maxContiguousHeight),
+			zap.Uint64("maxBlockHeight", heights.maxBlockHeight),
+		)
+	}
+	return nil
+}
+
+func (s *Database) recoverBlockAtOffset(offset, totalDataSize uint64) (blockEntryHeader, error) {
+	var bh blockEntryHeader
+	if totalDataSize-offset < uint64(sizeOfBlockEntryHeader) {
+		return bh, fmt.Errorf("%w: not enough data for block header at offset %d", ErrCorrupted, offset)
+	}
+
+	dataFile, localOffset, err := s.getDataFileAndOffset(offset)
+	if err != nil {
+		return bh, fmt.Errorf("recovery: failed to get data file for offset %d: %w", offset, err)
+	}
+	bhBuf := make([]byte, sizeOfBlockEntryHeader)
+	if _, err := dataFile.ReadAt(bhBuf, int64(localOffset)); err != nil {
+		return bh, fmt.Errorf("%w: error reading block header at offset %d: %w", ErrCorrupted, offset, err)
+	}
+	if err := bh.UnmarshalBinary(bhBuf); err != nil {
+		return bh, fmt.Errorf("%w: error deserializing block header at offset %d: %w", ErrCorrupted, offset, err)
+	}
+	if bh.Size == 0 {
+		return bh, fmt.Errorf("%w: invalid block size in header at offset %d: %d", ErrCorrupted, offset, bh.Size)
+	}
+	if bh.Version > BlockEntryVersion {
+		return bh, fmt.Errorf("%w: invalid block entry version at offset %d, version %d is greater than the current version %d", ErrCorrupted, offset, bh.Version, BlockEntryVersion)
+	}
+	if bh.Height < s.header.MinHeight || bh.Height == unsetHeight {
+		return bh, fmt.Errorf(
+			"%w: invalid block height in header at offset %d: found %d, expected >= %d",
+			ErrCorrupted, offset, bh.Height, s.header.MinHeight,
+		)
+	}
+	if bh.HeaderSize > bh.Size {
+		return bh, fmt.Errorf("%w: invalid block header size in header at offset %d: %d > %d", ErrCorrupted, offset, bh.HeaderSize, bh.Size)
+	}
+	expectedBlockEndOffset, err := safemath.Add(offset, uint64(sizeOfBlockEntryHeader))
+	if err != nil {
+		return bh, fmt.Errorf("calculating block end offset would overflow at offset %d: %w", offset, err)
+	}
+	expectedBlockEndOffset, err = safemath.Add(expectedBlockEndOffset, uint64(bh.Size))
+	if err != nil {
+		return bh, fmt.Errorf("calculating block end offset would overflow at offset %d: %w", offset, err)
+	}
+	if expectedBlockEndOffset > totalDataSize {
+		return bh, fmt.Errorf("%w: block data out of bounds at offset %d", ErrCorrupted, offset)
+	}
+	blockData := make([]byte, bh.Size)
+	blockDataOffset, err := safemath.Add(localOffset, uint64(sizeOfBlockEntryHeader))
+	if err != nil {
+		return bh, fmt.Errorf("calculating block data offset would overflow at offset %d: %w", offset, err)
+	}
+	if _, err := dataFile.ReadAt(blockData, int64(blockDataOffset)); err != nil {
+		return bh, fmt.Errorf("%w: failed to read block data at offset %d: %w", ErrCorrupted, offset, err)
+	}
+	calculatedChecksum := calculateChecksum(blockData)
+	if calculatedChecksum != bh.Checksum {
```

Comment:
If the checksum is wrong, why do we return error and fail the entire startup instead of truncating the block records from here after?

Reply:
yeah, I was referring to the case if that was the last record written.